File: backend/app/models/integration.py
"""
RFP Sniper - Integration Models
===============================
Stores integration configuration (SSO, CRM, storage, etc.).
"""

from datetime import datetime
from enum import Enum

from sqlmodel import JSON, Column, Field, SQLModel


class IntegrationProvider(str, Enum):
    OKTA = "okta"
    MICROSOFT = "microsoft"
    SHAREPOINT = "sharepoint"
    SALESFORCE = "salesforce"
    WORD_ADDIN = "word_addin"
    WEBHOOK = "webhook"
    SLACK = "slack"
    UNANET = "unanet"


class IntegrationSyncStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"


class IntegrationConfig(SQLModel, table=True):
    """
    Integration configuration record.
    """

    __tablename__ = "integrations"

    id: int | None = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="users.id", index=True)

    provider: IntegrationProvider = Field(index=True)
    name: str | None = Field(default=None, max_length=255)
    is_enabled: bool = Field(default=True)

    # Encrypted config stored as JSON (client secrets, endpoints, etc.)
    config: dict = Field(default={}, sa_column=Column(JSON))

    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class IntegrationSyncRun(SQLModel, table=True):
    """
    Integration sync history for external systems.
    """

    __tablename__ = "integration_sync_runs"

    id: int | None = Field(default=None, primary_key=True)
    integration_id: int = Field(foreign_key="integrations.id", index=True)
    provider: IntegrationProvider = Field(index=True)

    status: IntegrationSyncStatus = Field(default=IntegrationSyncStatus.PENDING)
    items_synced: int = Field(default=0)
    error: str | None = Field(default=None, max_length=1000)
    details: dict = Field(default={}, sa_column=Column(JSON))

    started_at: datetime = Field(default_factory=datetime.utcnow, index=True)
    completed_at: datetime | None = None


class IntegrationWebhookEvent(SQLModel, table=True):
    """
    Inbound webhook event payloads for integrations.
    """

    __tablename__ = "integration_webhook_events"

    id: int | None = Field(default=None, primary_key=True)
    integration_id: int = Field(foreign_key="integrations.id", index=True)
    provider: IntegrationProvider = Field(index=True)

    event_type: str = Field(default="generic", max_length=128, index=True)
    payload: dict = Field(default={}, sa_column=Column(JSON))

    received_at: datetime = Field(default_factory=datetime.utcnow, index=True)

File: frontend/src/types/integration.ts
// -----------------------------------------------------------------------------
// Integration Types
// -----------------------------------------------------------------------------

export type IntegrationProvider =
  | "okta"
  | "microsoft"
  | "sharepoint"
  | "salesforce"
  | "word_addin"
  | "webhook"
  | "slack";

export interface IntegrationConfig {
  id: number;
  provider: IntegrationProvider;
  name?: string | null;
  is_enabled: boolean;
  config: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface IntegrationFieldDefinition {
  key: string;
  label: string;
  secret: boolean;
  description?: string;
}

export interface IntegrationProviderDefinition {
  provider: IntegrationProvider;
  label: string;
  category: string;
  required_fields: IntegrationFieldDefinition[];
  optional_fields: IntegrationFieldDefinition[];
  supports_sync: boolean;
  supports_webhooks: boolean;
}

export interface IntegrationTestResult {
  status: "ok" | "error" | "disabled";
  message: string;
  missing_fields: string[];
  checked_at: string;
}

export interface IntegrationSyncRun {
  id: number;
  status: "pending" | "running" | "success" | "failed";
  items_synced: number;
  error?: string | null;
  details: Record<string, unknown>;
  started_at: string;
  completed_at?: string | null;
}

export interface IntegrationWebhookEvent {
  id: number;
  event_type: string;
  payload: Record<string, unknown>;
  received_at: string;
}

export interface IntegrationSsoAuthorizeResponse {
  provider: IntegrationProvider;
  authorization_url: string;
  state: string;
}

export interface AuditEvent {
  id: number;
  user_id?: number | null;
  entity_type: string;
  entity_id?: number | null;
  action: string;
  event_metadata: Record<string, unknown>;
  created_at: string;
}

export interface AuditSummary {
  period_days: number;
  total_events: number;
  by_action: Array<{ action: string; count: number }>;
  by_entity_type: Array<{ entity_type: string; count: number }>;
}

export interface ObservabilityMetrics {
  period_days: number;
  audit_events: { total: number };
  integration_syncs: {
    total: number;
    success: number;
    failed: number;
    last_sync_at?: string | null;
    by_provider: Record<string, { total: number; success: number; failed: number }>;
  };
  webhook_events: {
    total: number;
    by_provider: Record<string, number>;
  };
}
